File: 3.x.x/crates-utopia/openapi/src/parser/openapi2.rs
//! OpenAPI 2.0 reader (PHP `Parser\OpenAPI2`).

use crate::error::{InvalidSpecification, OpenApiError};
use crate::json::Json;
use crate::model::{
    Encoding, Example, HttpMethod, MediaType, ObjectSchema, Operation, Parameter,
    ParameterLocation, PathItem, RequestBody, Response, Schema, Server,
};
use crate::parser::reader::{json_strval, pointer_escape, DocumentReader};
use crate::parser::value::Value;
use crate::specification::Specification;
use indexmap::IndexMap;

pub fn read(ctx: &DocumentReader) -> Result<Specification, OpenApiError> {
    let security = ctx.parse_security(
        ctx.document
            .get("security")
            .unwrap_or(crate::json::empty_array()),
        "#/security",
    )?;
    let consumes = media_names(
        ctx.document
            .get("consumes")
            .unwrap_or(crate::json::empty_array()),
        "#/consumes",
    )?;
    let produces = media_names(
        ctx.document
            .get("produces")
            .unwrap_or(crate::json::empty_array()),
        "#/produces",
    )?;

    let mut schemas = IndexMap::new();
    for (name, raw) in Value::object(
        ctx.document
            .get("definitions")
            .unwrap_or(crate::json::empty_object()),
        "#/definitions",
    )? {
        schemas.insert(
            name.clone(),
            ctx.schemas.read(raw, &format!("#/definitions/{name}"))?,
        );
    }

    let mut security_schemes = IndexMap::new();
    for (name, raw) in Value::object(
        ctx.document
            .get("securityDefinitions")
            .unwrap_or(crate::json::empty_object()),
        "#/securityDefinitions",
    )? {
        let data = ctx.resolve_object(raw, &format!("#/securityDefinitions/{name}"))?;
        security_schemes.insert(
            name.clone(),
            ctx.parse_security_scheme(&data, &format!("#/securityDefinitions/{name}"), true)?,
        );
    }

    Ok(Specification {
        version: ctx.version,
        info: ctx.parse_info()?,
        servers: parse_open_api2_servers(ctx)?,
        tags: ctx.parse_tags()?,
        paths: parse_paths(ctx, &security, &consumes, &produces)?,
        schemas,
        security_schemes,
        security,
        extensions: Value::extensions(&ctx.document),
        source_version: ctx.source_version.clone(),
        json_schema_dialect: None,
        external_documentation: if ctx.document.contains_key("externalDocs") {
            Some(ctx.parse_external_documentation(
                ctx.document.get("externalDocs").unwrap_or(&Json::Null),
                "#/externalDocs",
            )?)
        } else {
            None
        },
    })
}

fn parse_open_api2_servers(ctx: &DocumentReader) -> Result<Vec<Server>, OpenApiError> {
    let host = Value::optional_string(&ctx.document, "host")?;
    let base_path = Value::optional_string(&ctx.document, "basePath")?.unwrap_or_default();
    let mut schemes = media_names(
        ctx.document
            .get("schemes")
            .unwrap_or(crate::json::empty_array()),
        "#/schemes",
    )?;
    let Some(host) = host else {
        return Ok(vec![]);
    };
    if schemes.is_empty() {
        schemes.push("http".into());
    }
    Ok(schemes
        .into_iter()
        .map(|scheme| {
            let mut url = format!("{scheme}://{host}");
            url.truncate(url.trim_end_matches('/').len());
            if base_path != "/" {
                url.push_str(&base_path);
            }
            Server::new(url)
        })
        .collect())
}

fn parse_paths(
    ctx: &DocumentReader,
    root_security: &[crate::model::SecurityRequirement],
    root_consumes: &[String],
    root_produces: &[String],
) -> Result<IndexMap<String, PathItem>, OpenApiError> {
    let mut paths = IndexMap::new();
    for (path, raw) in Value::object(
        ctx.document
            .get("paths")
            .unwrap_or(crate::json::empty_object()),
        "#/paths",
    )? {
        let location = format!("#/paths/{}", pointer_escape(path));
        let data = ctx.resolve_object(raw, &location)?;
        let path_parameters = raw_parameters(
            ctx,
            data.get("parameters").unwrap_or(crate::json::empty_array()),
            &format!("{location}/parameters"),
        )?;
        let mut operations = IndexMap::new();
        for method in HttpMethod::cases() {
            if !data.contains_key(method.as_str()) {
                continue;
            }
            operations.insert(
                method.as_str().to_owned(),
                parse_operation(
                    ctx,
                    path,
                    *method,
                    data.get(method.as_str()).unwrap_or(&Json::Null),
                    &path_parameters,
                    root_security,
                    root_consumes,
                    root_produces,
                    &format!("{location}/{}", method.as_str()),
                )?,
            );
        }
        let mut parsed_path_parameters = Vec::new();
        for (index, parameter) in path_parameters.iter().enumerate() {
            let in_ = parameter.get("in").and_then(Json::as_str);
            if in_ != Some("body") && in_ != Some("formData") {
                parsed_path_parameters.push(parse_parameter(
                    ctx,
                    parameter,
                    &format!("{location}/parameters/{index}"),
                )?);
            }
        }
        paths.insert(
            path.clone(),
            PathItem {
                path: path.clone(),
                operations,
                parameters: parsed_path_parameters,
                summary: String::new(),
                description: String::new(),
                servers: Vec::new(),
                extensions: Value::extensions(&data),
            },
        );
    }
    Ok(paths)
}

#[allow(clippy::too_many_arguments)]
fn parse_operation(
    ctx: &DocumentReader,
    path: &str,
    method: HttpMethod,
    raw: &Json,
    path_parameters: &[IndexMap<String, Json>],
    root_security: &[crate::model::SecurityRequirement],
    root_consumes: &[String],
    root_produces: &[String],
    location: &str,
) -> Result<Operation, OpenApiError> {
    let data = Value::object(raw, location)?;
    let operation_parameters = raw_parameters(
        ctx,
        data.get("parameters").unwrap_or(crate::json::empty_array()),
        &format!("{location}/parameters"),
    )?;
    let raw_parameters = merge_raw_parameters(path_parameters, &operation_parameters);
    let consumes = if data.contains_key("consumes") {
        media_names(
            data.get("consumes").unwrap_or(&Json::Null),
            &format!("{location}/consumes"),
        )?
    } else {
        root_consumes.to_vec()
    };
    let produces = if data.contains_key("produces") {
        media_names(
            data.get("produces").unwrap_or(&Json::Null),
            &format!("{location}/produces"),
        )?
    } else {
        root_produces.to_vec()
    };

    let mut parameters = Vec::new();
    let mut body = None;
    let mut form = Vec::new();
    for (index, parameter) in raw_parameters.iter().enumerate() {
        let in_ = parameter.get("in").and_then(Json::as_str);
        if in_ == Some("body") {
            if body.is_some() {
                return Err(InvalidSpecification(format!(
                    "Multiple body parameters at {location}"
                ))
                .into());
            }
            let schema = if parameter.contains_key("schema") {
                Some(ctx.schemas.read(
                    parameter.get("schema").unwrap_or(&Json::Null),
                    &format!("{location}/parameters/{index}/schema"),
                )?)
            } else {
                None
            };
            let mut content = IndexMap::new();
            for media_name in &consumes {
                content.insert(media_name.clone(), MediaType::new(schema.clone()));
            }
            body = Some(RequestBody {
                description: Value::optional_string(parameter, "description")?.unwrap_or_default(),
                required: parameter.get("required").is_some_and(Json::php_bool),
                content,
                extensions: Value::extensions(parameter),
            });
        } else if in_ == Some("formData") {
            form.push(parameter);
        } else {
            parameters.push(parse_parameter(
                ctx,
                parameter,
                &format!("{location}/parameters/{index}"),
            )?);
        }
    }
    if body.is_some() && !form.is_empty() {
        return Err(InvalidSpecification(format!(
            "Body and formData parameters cannot coexist at {location}"
        ))
        .into());
    }
    if !form.is_empty() {
        body = Some(parse_form_body(ctx, &form, &consumes, location)?);
    }

    let mut responses = IndexMap::new();
    for (status, raw_response) in Value::object(
        data.get("responses").unwrap_or(&Json::Null),
        &format!("{location}/responses"),
    )? {
        let value = ctx.resolve_object(raw_response, &format!("{location}/responses/{status}"))?;
        let examples_by_media = open_api2_examples(
            value.get("examples").unwrap_or(crate::json::empty_object()),
            &format!("{location}/responses/{status}/examples"),
        )?;
        let schema = if value.contains_key("schema") {
            Some(ctx.schemas.read(
                value.get("schema").unwrap_or(&Json::Null),
                &format!("{location}/responses/{status}/schema"),
            )?)
        } else {
            None
        };
        let mut media_names: Vec<String> = produces.clone();
        for key in examples_by_media.keys() {
            if !media_names.contains(key) {
                media_names.push(key.clone());
            }
        }
        let mut content = IndexMap::new();
        for media_name in media_names {
            let example = examples_by_media
                .get(&media_name)
                .map_or(Json::Null, |e| e.value.clone());
            content.insert(
                media_name,
                MediaType {
                    schema: schema.clone(),
                    example,
                    examples: IndexMap::new(),
                    encoding: IndexMap::new(),
                    extensions: IndexMap::new(),
                },
            );
        }
        responses.insert(
            status.clone(),
            Response {
                description: Value::required_string(
                    &value,
                    "description",
                    &format!("{location}/responses/{status}"),
                )?,
                headers: ctx.parse_headers(
                    value.get("headers").unwrap_or(crate::json::empty_object()),
                    &format!("{location}/responses/{status}/headers"),
                    true,
                )?,
                content,
                extensions: Value::extensions(&value),
            },
        );
    }

    Ok(Operation {
        id: Value::optional_string(data, "operationId")?.unwrap_or_default(),
        method,
        path: path.to_owned(),
        tags: if data.contains_key("tags") {
            Value::list(
                data.get("tags").unwrap_or(crate::json::empty_array()),
                &format!("{location}/tags"),
            )?
            .iter()
            .map(json_strval)
            .collect()
        } else {
            Vec::new()
        },
        summary: Value::optional_string(data, "summary")?.unwrap_or_default(),
        description: Value::optional_string(data, "description")?.unwrap_or_default(),
        deprecated: data.get("deprecated").is_some_and(Json::php_bool),
        parameters,
        request_body: body,
        responses,
        security: if data.contains_key("security") {
            ctx.parse_security(
                data.get("security").unwrap_or(&Json::Null),
                &format!("{location}/security"),
            )?
        } else {
            root_security.to_vec()
        },
        servers: Vec::new(),
        external_documentation: if data.contains_key("externalDocs") {
            Some(ctx.parse_external_documentation(
                data.get("externalDocs").unwrap_or(&Json::Null),
                &format!("{location}/externalDocs"),
            )?)
        } else {
            None
        },
        extensions: Value::extensions(data),
    })
}

fn raw_parameters(
    ctx: &DocumentReader,
    raw: &Json,
    location: &str,
) -> Result<Vec<IndexMap<String, Json>>, OpenApiError> {
    let mut parameters = Vec::new();
    for (index, item) in Value::list(raw, location)?.iter().enumerate() {
        parameters.push(ctx.resolve_object(item, &format!("{location}/{index}"))?);
    }
    Ok(parameters)
}

fn parse_parameter(
    ctx: &DocumentReader,
    data: &IndexMap<String, Json>,
    location: &str,
) -> Result<Parameter, OpenApiError> {
    let location_value = Value::required_string(data, "in", location)?;
    let parameter_location = ParameterLocation::from_str_php(&location_value).ok_or_else(|| {
        InvalidSpecification(format!(
            "Unsupported parameter location '{location_value}' at {location}/in"
        ))
    })?;
    let required = data.get("required").is_some_and(Json::php_bool);
    if parameter_location == ParameterLocation::Path && !required {
        return Err(
            InvalidSpecification(format!("Path parameter must be required at {location}")).into(),
        );
    }
    Ok(Parameter {
        name: Value::required_string(data, "name", location)?,
        location: parameter_location,
        description: Value::optional_string(data, "description")?.unwrap_or_default(),
        required,
        deprecated: false,
        allow_empty_value: data.get("allowEmptyValue").is_some_and(Json::php_bool),
        schema: Some(ctx.schemas.read_parameter_fields(data, location)?),
        content: IndexMap::new(),
        style: None,
        explode: None,
        allow_reserved: false,
        extensions: Value::extensions(data),
    })
}

fn parse_form_body(
    ctx: &DocumentReader,
    form: &[&IndexMap<String, Json>],
    consumes: &[String],
    location: &str,
) -> Result<RequestBody, OpenApiError> {
    let mut properties = IndexMap::new();
    let mut required = Vec::new();
    let mut encoding = IndexMap::new();
    for (index, parameter) in form.iter().enumerate() {
        let name =
            Value::required_string(parameter, "name", &format!("{location}/parameters/{index}"))?;
        properties.insert(
            name.clone(),
            ctx.schemas
                .read_parameter_fields(parameter, &format!("{location}/parameters/{index}"))?,
        );
        if parameter.get("required").is_some_and(Json::php_bool) {
            required.push(name.clone());
        }
        encoding.insert(
            name,
            Encoding {
                content_type: if parameter.get("type").and_then(Json::as_str) == Some("file") {
                    Some("application/octet-stream".into())
                } else {
                    None
                },
                extensions: Value::extensions(parameter),
                ..Encoding::default()
            },
        );
    }
    let schema = Schema::Object(Box::new(ObjectSchema {
        meta: crate::model::SchemaMeta::default(),
        properties,
        required: required.clone(),
        additional_properties: None,
        min_properties: None,
        max_properties: None,
    }));
    let consumes = if consumes.is_empty() {
        vec!["application/x-www-form-urlencoded".to_owned()]
    } else {
        consumes.to_vec()
    };
    let mut content = IndexMap::new();
    for media_name in consumes {
        content.insert(
            media_name,
            MediaType {
                schema: Some(schema.clone()),
                example: Json::Null,
                examples: IndexMap::new(),
                encoding: encoding.clone(),
                extensions: IndexMap::new(),
            },
        );
    }
    Ok(RequestBody {
        description: String::new(),
        required: !required.is_empty(),
        content,
        extensions: IndexMap::new(),
    })
}

fn open_api2_examples(
    raw: &Json,
    location: &str,
) -> Result<IndexMap<String, Example>, OpenApiError> {
    let mut examples = IndexMap::new();
    for (media_name, value) in Value::object(raw, location)? {
        examples.insert(
            media_name.clone(),
            Example {
                value: value.clone(),
                ..Example::default()
            },
        );
    }
    Ok(examples)
}

fn merge_raw_parameters(
    inherited: &[IndexMap<String, Json>],
    operation: &[IndexMap<String, Json>],
) -> Vec<IndexMap<String, Json>> {
    let mut merged = inherited.to_vec();
    let mut indexes = IndexMap::new();
    for (index, parameter) in merged.iter().enumerate() {
        indexes.insert(raw_parameter_identity(parameter), index);
    }
    for parameter in operation {
        let identity = raw_parameter_identity(parameter);
        if let Some(&idx) = indexes.get(&identity) {
            merged[idx].clone_from(parameter);
        } else {
            indexes.insert(identity, merged.len());
            merged.push(parameter.clone());
        }
    }
    merged
}

fn raw_parameter_identity(parameter: &IndexMap<String, Json>) -> String {
    let in_ = parameter.get("in").map(json_strval).unwrap_or_default();
    let name = parameter.get("name").map(json_strval).unwrap_or_default();
    format!("{in_}\0{name}")
}

fn media_names(raw: &Json, location: &str) -> Result<Vec<String>, OpenApiError> {
    Ok(Value::list(raw, location)?
        .iter()
        .map(json_strval)
        .collect())
}

File: packages/openapi/src/Parser/OpenAPI2.php
<?php

declare(strict_types=1);

namespace Utopia\OpenAPI\Parser;

use Utopia\OpenAPI\Exception\InvalidSpecification;
use Utopia\OpenAPI\Model\Encoding;
use Utopia\OpenAPI\Model\Example;
use Utopia\OpenAPI\Model\HttpMethod;
use Utopia\OpenAPI\Model\MediaType;
use Utopia\OpenAPI\Model\ObjectSchema;
use Utopia\OpenAPI\Model\Operation;
use Utopia\OpenAPI\Model\Parameter;
use Utopia\OpenAPI\Model\ParameterLocation;
use Utopia\OpenAPI\Model\PathItem;
use Utopia\OpenAPI\Model\RequestBody;
use Utopia\OpenAPI\Model\Response;
use Utopia\OpenAPI\Model\Server;
use Utopia\OpenAPI\Specification;

final class OpenAPI2 extends AbstractReader
{
    #[\Override]
    public function read(): Specification
    {
        $security = $this->parseSecurity($this->document['security'] ?? [], '#/security');
        $consumes = $this->mediaNames($this->document['consumes'] ?? [], '#/consumes');
        $produces = $this->mediaNames($this->document['produces'] ?? [], '#/produces');

        $schemas = [];
        foreach (Value::object($this->document['definitions'] ?? [], '#/definitions') as $name => $raw) {
            $schemas[(string) $name] = $this->schemas->read($raw, "#/definitions/{$name}");
        }

        $securitySchemes = [];
        foreach (Value::object($this->document['securityDefinitions'] ?? [], '#/securityDefinitions') as $name => $raw) {
            $data = $this->resolveObject($raw, "#/securityDefinitions/{$name}");
            $securitySchemes[(string) $name] = $this->parseSecurityScheme($data, "#/securityDefinitions/{$name}", true);
        }

        return new Specification(
            version: $this->version,
            info: $this->parseInfo(),
            servers: $this->parseOpenApi2Servers(),
            tags: $this->parseTags(),
            paths: $this->parsePaths($security, $consumes, $produces),
            schemas: $schemas,
            securitySchemes: $securitySchemes,
            security: $security,
            extensions: Value::extensions($this->document),
            sourceVersion: $this->sourceVersion,
            externalDocumentation: isset($this->document['externalDocs'])
                ? $this->parseExternalDocumentation($this->document['externalDocs'], '#/externalDocs')
                : null,
        );
    }

    /** @return list<Server> */
    private function parseOpenApi2Servers(): array
    {
        $host = Value::optionalString($this->document, 'host');
        $basePath = Value::optionalString($this->document, 'basePath') ?? '';
        $schemes = $this->mediaNames($this->document['schemes'] ?? [], '#/schemes');
        if ($host === null) {
            return [];
        }
        if ($schemes === []) {
            $schemes = ['http'];
        }

        return array_map(
            static fn(string $scheme): Server => new Server(rtrim($scheme . '://' . $host, '/') . ($basePath === '/' ? '' : $basePath)),
            $schemes,
        );
    }

    /** @return array<string, PathItem> */
    private function parsePaths(array $rootSecurity, array $rootConsumes, array $rootProduces): array
    {
        $paths = [];
        foreach (Value::object($this->document['paths'] ?? [], '#/paths') as $path => $raw) {
            $location = '#/paths/' . str_replace(['~', '/'], ['~0', '~1'], (string) $path);
            $data = $this->resolveObject($raw, $location);
            $pathParameters = $this->rawParameters($data['parameters'] ?? [], "{$location}/parameters");
            $operations = [];
            foreach (HttpMethod::cases() as $method) {
                if (! \array_key_exists($method->value, $data)) {
                    continue;
                }
                $operations[$method->value] = $this->parseOperation((string) $path, $method, $data[$method->value], $pathParameters, $rootSecurity, $rootConsumes, $rootProduces, "{$location}/{$method->value}");
            }
            $parsedPathParameters = [];
            foreach ($pathParameters as $index => $parameter) {
                if (($parameter['in'] ?? null) !== 'body' && ($parameter['in'] ?? null) !== 'formData') {
                    $parsedPathParameters[] = $this->parseParameter($parameter, "{$location}/parameters/{$index}");
                }
            }
            $paths[(string) $path] = new PathItem((string) $path, $operations, $parsedPathParameters, extensions: Value::extensions($data));
        }

        return $paths;
    }

    private function parseOperation(string $path, HttpMethod $method, mixed $raw, array $pathParameters, array $rootSecurity, array $rootConsumes, array $rootProduces, string $location): Operation
    {
        $data = Value::object($raw, $location);
        $operationParameters = $this->rawParameters($data['parameters'] ?? [], "{$location}/parameters");
        $rawParameters = $this->mergeRawParameters($pathParameters, $operationParameters);
        $consumes = \array_key_exists('consumes', $data) ? $this->mediaNames($data['consumes'], "{$location}/consumes") : $rootConsumes;
        $produces = \array_key_exists('produces', $data) ? $this->mediaNames($data['produces'], "{$location}/produces") : $rootProduces;

        $parameters = [];
        $body = null;
        $form = [];
        foreach ($rawParameters as $index => $parameter) {
            $in = $parameter['in'] ?? null;
            if ($in === 'body') {
                if ($body !== null) {
                    throw new InvalidSpecification("Multiple body parameters at {$location}");
                }
                $schema = \array_key_exists('schema', $parameter) ? $this->schemas->read($parameter['schema'], "{$location}/parameters/{$index}/schema") : null;
                $content = [];
                foreach ($consumes as $mediaName) {
                    $content[$mediaName] = new MediaType($schema);
                }
                $body = new RequestBody(
                    description: Value::optionalString($parameter, 'description') ?? '',
                    required: (bool) ($parameter['required'] ?? false),
                    content: $content,
                    extensions: Value::extensions($parameter),
                );
            } elseif ($in === 'formData') {
                $form[] = $parameter;
            } else {
                $parameters[] = $this->parseParameter($parameter, "{$location}/parameters/{$index}");
            }
        }
        if ($body !== null && $form !== []) {
            throw new InvalidSpecification("Body and formData parameters cannot coexist at {$location}");
        }
        if ($form !== []) {
            $body = $this->parseFormBody($form, $consumes, $location);
        }

        $responses = [];
        foreach (Value::object($data['responses'] ?? null, "{$location}/responses") as $status => $rawResponse) {
            $value = $this->resolveObject($rawResponse, "{$location}/responses/{$status}");
            $content = [];
            $examplesByMedia = $this->openApi2Examples($value['examples'] ?? [], "{$location}/responses/{$status}/examples");
            $schema = \array_key_exists('schema', $value)
                ? $this->schemas->read($value['schema'], "{$location}/responses/{$status}/schema")
                : null;
            foreach (array_values(array_unique([...$produces, ...array_keys($examplesByMedia)])) as $mediaName) {
                $content[$mediaName] = new MediaType(
                    schema: $schema,
                    example: $examplesByMedia[$mediaName]->value ?? null,
                );
            }
            $responses[(string) $status] = new Response(
                description: Value::requiredString($value, 'description', "{$location}/responses/{$status}"),
                headers: $this->parseHeaders($value['headers'] ?? [], "{$location}/responses/{$status}/headers", true),
                content: $content,
                extensions: Value::extensions($value),
            );
        }

        return new Operation(
            id: Value::optionalString($data, 'operationId') ?? '',
            method: $method,
            path: $path,
            tags: isset($data['tags']) ? array_map(strval(...), Value::list($data['tags'], "{$location}/tags")) : [],
            summary: Value::optionalString($data, 'summary') ?? '',
            description: Value::optionalString($data, 'description') ?? '',
            deprecated: (bool) ($data['deprecated'] ?? false),
            parameters: $parameters,
            requestBody: $body,
            responses: $responses,
            security: \array_key_exists('security', $data) ? $this->parseSecurity($data['security'], "{$location}/security") : $rootSecurity,
            externalDocumentation: isset($data['externalDocs']) ? $this->parseExternalDocumentation($data['externalDocs'], "{$location}/externalDocs") : null,
            extensions: Value::extensions($data),
        );
    }

    /** @return list<array<string, mixed>> */
    private function rawParameters(mixed $raw, string $location): array
    {
        $parameters = [];
        foreach (Value::list($raw, $location) as $index => $item) {
            $parameters[] = $this->resolveObject($item, "{$location}/{$index}");
        }

        return $parameters;
    }

    private function parseParameter(array $data, string $location): Parameter
    {
        $locationValue = Value::requiredString($data, 'in', $location);
        try {
            $parameterLocation = ParameterLocation::from($locationValue);
        } catch (\ValueError) {
            throw new InvalidSpecification("Unsupported parameter location '{$locationValue}' at {$location}/in");
        }
        $required = (bool) ($data['required'] ?? false);
        if ($parameterLocation === ParameterLocation::PATH && ! $required) {
            throw new InvalidSpecification("Path parameter must be required at {$location}");
        }

        return new Parameter(
            name: Value::requiredString($data, 'name', $location),
            location: $parameterLocation,
            description: Value::optionalString($data, 'description') ?? '',
            required: $required,
            allowEmptyValue: (bool) ($data['allowEmptyValue'] ?? false),
            schema: $this->schemas->readParameterFields($data, $location),
            extensions: Value::extensions($data),
        );
    }

    private function parseFormBody(array $form, array $consumes, string $location): RequestBody
    {
        $properties = [];
        $required = [];
        $encoding = [];
        foreach ($form as $index => $parameter) {
            $name = Value::requiredString($parameter, 'name', "{$location}/parameters/{$index}");
            $properties[$name] = $this->schemas->readParameterFields($parameter, "{$location}/parameters/{$index}");
            if ((bool) ($parameter['required'] ?? false)) {
                $required[] = $name;
            }
            $encoding[$name] = new Encoding(
                contentType: ($parameter['type'] ?? null) === 'file' ? 'application/octet-stream' : null,
                extensions: Value::extensions($parameter),
            );
        }
        $schema = new ObjectSchema($properties, $required);
        if ($consumes === []) {
            $consumes = ['application/x-www-form-urlencoded'];
        }
        $content = [];
        foreach ($consumes as $mediaName) {
            $content[$mediaName] = new MediaType($schema, encoding: $encoding);
        }

        return new RequestBody(required: $required !== [], content: $content);
    }

    /** @return array<string, Example> */
    private function openApi2Examples(mixed $raw, string $location): array
    {
        $examples = [];
        foreach (Value::object($raw, $location) as $mediaName => $value) {
            $examples[(string) $mediaName] = new Example(value: $value);
        }

        return $examples;
    }

    private function mergeRawParameters(array $inherited, array $operation): array
    {
        $merged = $inherited;
        $indexes = [];
        foreach ($merged as $index => $parameter) {
            $indexes[$this->rawParameterIdentity($parameter)] = $index;
        }
        foreach ($operation as $parameter) {
            $identity = $this->rawParameterIdentity($parameter);
            if (\array_key_exists($identity, $indexes)) {
                $merged[$indexes[$identity]] = $parameter;
            } else {
                $indexes[$identity] = \count($merged);
                $merged[] = $parameter;
            }
        }

        return array_values($merged);
    }

    private function rawParameterIdentity(array $parameter): string
    {
        return (string) ($parameter['in'] ?? '') . "\0" . (string) ($parameter['name'] ?? '');
    }

    /** @return list<string> */
    private function mediaNames(mixed $raw, string $location): array
    {
        return array_map(strval(...), Value::list($raw, $location));
    }
}
